### DISCUSSION TITLE
Slingbox + GerryDazoo/Slinger + Xteve + Plex Live TV

### DISCUSSION THREAD

**Discussion by rbf121** [GerryDazoo/Slinger#79] [Ideas]
Edit 11/30/2022:  Plex Live TV integration with Slinger has been running well for me for several weeks.  Find the most up to date setup instructions here:  https://www.reddit.com/r/slingbox/comments/xhy710/plex_live_tv_integration_with_slingbox/

Also possible to set this up with Jellyfin using just the M3U file:
https://www.reddit.com/r/slingbox/comments/yttc4u/slingbox_with_jellyfin_live_tv/

**Comment by rbf121** [comment]
I modified some of the python code locally to change the channel if it's passed into the stream and have it working in Plex.  The m3u file lists out all of the channels generically, ie Channel 1 through 160.  Then in Plex these get mapped automatically to the correct EPG data.  Plex shows it all in a nice grid view with upcoming programming.

You can select a channel in Plex via grid view which loads the stream defined in the m3u file (in this case, it would be http://ipaddress:port/slingbox?channel=20).   This loads the slingbox stream and kicks off a thread to submit the IR signals to switch the channel.

I still need to figure out how to get Plex to stop timing out on loading the stream.  Maybe if I could get the slinger python app to play the correct format then the ffmpeg buffering won't be necessary and plex won't time out.

**Reply by GerryDazoo** [reply]
The slinger code does not do any video encoding/decoding. What you get is what comes straight out of the slingbox. Basically all it does is setup the stream and then just bit bashes packet in - packet out. Very low overhead. Adding this functionally is way out of scope for what this code was originally intended. But if you want to take it as a base and bend it to your will, you're welcome to it. I suppose it would be possible to pass the stream from my program through a ffmpeg post processor to get what you want.

**Reply by rbf121** [reply]
Understood.  Xteve is actually doing the ffmpeg post processing part so should be okay.  I tweaked some of the Xteve buffer settings and it's working reasonably well now.  No longer timing out in plex.

Through Plex, I have guide data automatically populated, I can change the channels based on selecting a channel from the guide and there is an option to record the show as well.  In addition, it should allow plex to handle all of the port forwarding etc to watch it from outside your network.

Thanks for your work on this project and sharing the code.

**Reply by tom76017** [reply]
Looking forward to testing some beta code.   I'm running both Plex and Slinger on an older HP Compaq Elite 8000 with Debian 11.5.  There are no CPU issues with Slinger, just I/O.  Will this solution be packaged as a Plex Plugin?

**Reply by rbf121** [reply]
> Looking forward to testing some beta code. I'm running both Plex and Slinger on an older HP Compaq Elite 8000 with Debian 11.5. There are no CPU issues with Slinger, just I/O. Will this solution be packaged as a Plex Plugin?

No, it's not a plex plugin.  It's another piece of software that creates a TV Tuner which Plex can integrate into.  I have Plex, xTeVe, and Slinger running on the same server with no issues.  I'll work on a write up and post it.

**Reply by rbf121** [reply]
I wrote something up here - https://www.reddit.com/r/slingbox/comments/xhy710/plex_live_tv_integration_with_slingbox/

**Reply by GerryDazoo** [reply]
Thanks for the info.

**Reply by GerryDazoo** [reply]
I've integrated this capability ?channel=xxx into the Slinger mainline. You'll see it in the 3.08 release coming soon.

**Reply by rbf121** [reply]
Great, thanks!  You may want to review the remotes[port] section of the method call I created.  I copied the IR code from somewhere lower in the python file and I'm not entirely sure this will work in every scenario.  It worked in my use case so I didn't investigate more but I didn't need to tweak any of my remote settings either.

**Reply by GerryDazoo** [reply]
I ended up using your idea but not implementation. I parse the URL in the connection_manager and pass it to the streamer thread on it's message queue. With your separate thread there was an interaction with my new "LockRemote" feature and also there was a race condition with the Solo stream startup.

**Reply by rbf121** [reply]
Nice, I'll have to try it out when it's released.  I thought about that race condition where I might need to add a sleep but it just worked on my M1 so I left it alone to have faster channel changing.

**Comment by rbf121** [comment]
If anyone is interested, I uploaded the sample files for Xteve that I used to integrate the Slinger software with Plex.  I also included a Dockerfile which will set up a container to run both Xteve and Slinger in the same container, including all the dependencies needed.  This is working on my NAS but would be curious if anyone else gets it to work correctly.

It's over in my fork https://github.com/rbf121/Slinger

**Reply by flipper2222222** [reply]
thanks to rbf121 I am making some progress on configuring for windows. will update once I get it going.

**Reply by Wildwillis** [reply]
Thanks for creating the fork and posting your Dockerfile. I’d like to try and get the same config you have running on my QNAP NAS. I am quite a novice with Docker so I was hoping to get some help understanding how to properly mount the Volumes.

I have run a few Dockerfiles like the one you posted, but they’ve all specified directory paths within the Dockerfile. I typically start from something like that, modify any options I need to change and the directory paths, then SSH into my NAS and paste in the whole Dockerfile.

I see how you have a Volumes section in the Dockerfile instead. If I’m understanding your ReadMe correctly, I should mount these Volumes with the directories I need before running the Dockerfile. I am not entirely sure how to do that.

Any chance you can help with that or provide an example. I’ve tried to look into this a bit and I understand that mounting the volumes is persistent. I think that means that I only need to mount them once - if I update or recreate the Docker file to make changes, I don’t have to re-mount the directories again. Please confirm if I’m making myself clear. Thanks so much again!

**Comment by rbf121** [comment]
I decided to run xteve in it's own docker container because I needed 2 instances of xteve to meet my needs and it just makes more sense to have different containers.  I have updated the code in my fork (https://github.com/rbf121/Slinger) to include a Dockerfile that just runs GerryDazoo/Slinger latest code, v3.0.8 plus I included my samples for xteve and m3u in case anyone needs those.

I also published an image on docker hub but I need to still work on the instructions there.  I have this running on my NAS in host mode but I'm just wondering if anyone else can get it to work.  Let me know if there are any issues.
https://hub.docker.com/r/rbf121/slinger

I'm currently running the following for the xteve/slinger/plex integration:

- 2 x https://hub.docker.com/r/scp002/xteve  (this version of xteve seems more updated than https://hub.docker.com/r/alturismo/xteve)
- https://hub.docker.com/r/rbf121/slinger

**Comment by pfuller88** [comment]
I'm having trouble getting this to work.  I think I'm 99% there.   I follow all the instructions and it plex seems to be connecting via FFMPEG but plex times out with the Playback error, Could not tune channel, please check your tuner or antenna.   My xteve log shows the following.  Any assistance is appreicated.   I have xteve running natively on an Ubuntu 22 server - as is Slinger (same server).  I use a unified config file for slinger with the /condo/500high address, which seems to show correctly below.  The Plex server was able to build my proper TV guide.  Just cannot stream.   I did use the suggested FFMPEG setting:  -hide_banner -loglevel error -fflags +genpts -i [URL] -c copy -bsf:v h264_metadata=sample_aspect_ratio=4/3 -f mpegts -tune zerolatency pipe:1?

ubuntu log...


022/11/18 20:36:12 [xTeVe] XEPG:                   Ready to use
2022/11/18 20:37:22 [xTeVe] Buffer:                 true [ffmpeg]
2022/11/18 20:37:22 [xTeVe] Buffer Size:            1024 KB
2022/11/18 20:37:22 [xTeVe] Channel Name:           Slingbox 806
2022/11/18 20:37:22 [xTeVe] Client User-Agent:      Lavf/58.65.101
2022/11/18 20:37:23 [xTeVe] Streaming Status:       Playlist: M3U - Tuner: 1 / 1
2022/11/18 20:37:23 [xTeVe] FFMPEG path:            /usr/bin/ffmpeg
2022/11/18 20:37:23 [xTeVe] Streaming URL:          http://127.0.0.1:8080/condo/500high?channel=806
2022/11/18 20:37:23 [xTeVe] FFMPEG:                 Processing data
2022/11/18 20:37:23 [xTeVe] Streaming Status:       Receive data from FFMPEG
2022/11/18 20:37:43 [xTeVe] Streaming Status:       Client has terminated the connection
2022/11/18 20:37:43 [xTeVe] Streaming Status:       Channel: Slingbox 806 (Clients: 0)
2022/11/18 20:37:45 [xTeVe] Streaming Status:       Channel: Slingbox 806 - No client is using this channel anymore. Streaming Server connection has ended
2022/11/18 20:37:45 [xTeVe] Streaming Status:       Playlist: M3U - Tuner: 0 / 1

**Reply by pfuller88** [reply]
I did some more testing, Plex is changing the channel ( I can connect via VLC and see the channel has changed to what I choose in teh Plex guide) so it is connecting, just does not steam the video.

**Reply by rbf121** [reply]
Only difference I can see is the buffer size you have 1024 KB, try dropping that to 512 KB.

You also may not need the -bsf:v h264_metadata=sample_aspect_ratio=4/3 part, I have since removed it from the base instructions on reddit since it seemed to only be a weird aspect ratio issue I encountered.

**Reply by pfuller88** [reply]
Thanks for your reply.  I'm stumped on this one.  I change the FFPMEG settings and the bufffer, have tried a lot.  I have moved to the windows version of xteve and ffmpeg.  Still will not stream in Plex.  Everything seems operational, plex looks like it's connecting and then trying to trasncode which stops the process - i.e. from below, client has terminated the connection?  Now all plex does is quick back to the guide without a message.  Here is the xteve log.  Any more thoughts are apprecaited.

2022/11/19 09:29:46 [xTeVe] All streams:            22
2022/11/19 09:29:46 [xTeVe] Active streams:         22
2022/11/19 09:29:46 [xTeVe] Filter:                 0
2022/11/19 09:29:46 [xTeVe] XEPG:                   Parse XMLTV file: tivo
2022/11/19 09:29:46 [xTeVe] XEPG:                   Update database
2022/11/19 09:29:46 [xTeVe] XEPG:                   Save DB file
2022/11/19 09:29:46 [xTeVe] XEPG:                   Map channels
2022/11/19 09:29:46 [xTeVe] XEPG:                   Cleanup database
2022/11/19 09:29:46 [xTeVe] XEPG Channels:          22
2022/11/19 09:29:46 [xTeVe] DVR IP:                 #######:34400
2022/11/19 09:29:46 [xTeVe] Web Interface:          http://#######:34400/web/ | xTeVe is also available via the other 7 IP's.
2022/11/19 09:29:46 [xTeVe] XEPG:                   Create XMLTV file (C:\Users\pbulk\.xteve\data\xteve.xml)
2022/11/19 09:29:46 [xTeVe] XEPG:                   Compress XMLTV file (C:\Users\pbulk\.xteve\data\xteve.xml.gz)
2022/11/19 09:29:46 [xTeVe] XEPG:                   Create M3U file (C:\Users\pbulk\.xteve\data\xteve.m3u)
2022/11/19 09:29:46 [xTeVe] XEPG:                   Ready to use
2022/11/19 09:31:59 [xTeVe] Buffer:                 true [ffmpeg]
2022/11/19 09:31:59 [xTeVe] Buffer Size:            512 KB
2022/11/19 09:31:59 [xTeVe] Channel Name:           BBC America 876
2022/11/19 09:31:59 [xTeVe] Client User-Agent:      Lavf/LIBAVFORMAT_VERSION
2022/11/19 09:32:00 [xTeVe] Streaming Status:       Playlist: tivo - Tuner: 1 / 1
2022/11/19 09:32:00 [xTeVe] FFMPEG path:            C:\ffmpeg\bin\ffmpeg.exe
2022/11/19 09:32:00 [xTeVe] Streaming URL:          http://192.168.0.50:8180/condo/500med?channel=876
2022/11/19 09:32:00 [xTeVe] FFMPEG:                 Processing data
2022/11/19 09:32:00 [xTeVe] Streaming Status:       Receive data from FFMPEG
2022/11/19 09:32:00 [xTeVe] FFMPEG log:             ffmpeg version 2022-11-03-git-5ccd4d3060-full_build-www.gyan.dev Copyright (c) 2000-2022 the FFmpeg developers
2022/11/19 09:32:00 [xTeVe] FFMPEG log:             built with gcc 12.1.0 (Rev2, Built by MSYS2 project)
2022/11/19 09:32:00 [xTeVe] [ERROR] FFMPEG error (Streaming was stopped by third party transcoder (FFmpeg / VLC)) - EC: 1204
2022/11/19 09:32:00 [xTeVe] FFMPEG log:             configuration: --enable-gpl --enable-version3 --enable-static --disable-w32threads --disable-autodetect --enable-fontconfig --enable-iconv --enable-gnutls --enable-libxml2 --enable-gmp --enable-bzlib --enable-lzma --enable-libsnappy --enable-zlib --enable-librist --enable-libsrt --enable-libssh --enable-libzmq --enable-avisynth --enable-libbluray --enable-libcaca --enable-sdl2 --enable-libaribb24 --enable-libdav1d --enable-libdavs2 --enable-libuavs3d --enable-libzvbi --enable-librav1e --enable-libsvtav1 --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libaom --enable-libjxl --enable-libopenjpeg --enable-libvpx --enable-mediafoundation --enable-libass --enable-frei0r --enable-libfreetype --enable-libfribidi --enable-liblensfun --enable-libvidstab --enable-libvmaf --enable-libzimg --enable-amf --enable-cuda-llvm --enable-cuvid --enable-ffnvcodec --enable-nvdec --enable-nvenc --enable-d3d11va --enable-dxva2 --enable-libvpl --enable-libshaderc --enable-vulkan --enable-libplacebo --enable-opencl --enable-libcdio --enable-libgme --enable-libmodplug --enable-libopenmpt --enable-libopencore-amrwb --enable-libmp3lame --enable-libshine --enable-libtheora --enable-libtwolame --enable-libvo-amrwbenc --enable-libilbc --enable-libgsm --enable-libopencore-amrnb --enable-libopus --enable-libspeex --enable-libvorbis --enable-ladspa --enable-libbs2b --enable-libflite --enable-libmysofa --enable-librubberband --enable-libsoxr --enable-chromaprint
2022/11/19 09:32:00 [xTeVe] Streaming Status:       Client has terminated the connection
2022/11/19 09:32:00 [xTeVe] Streaming Status:       Channel: BBC America 876 (Clients: 0)
2022/11/19 09:32:01 [xTeVe] Streaming Status:       Channel: BBC America 876 - No client is using this channel anymore. Streaming Server connection has ended
2022/11/19 09:32:01 [xTeVe] Streaming Status:       Playlist: tivo - Tuner: 0 / 1

**Reply by GerryDazoo** [reply]
Here's the runtime parameter I use for ffplayer to minimize network delays due to buffering
C:\ffmpeg\ffplay.exe -infbuf -probesize 32 -analyzeduration 0 -vf scale=1280x720 -flags nobuffer -flags low_delay
Please post the output from slingbox_server

PS. This is the ffplayer version I'm using. ffplay version 2022-05-04-git-0914e3a14a-essentials_build-www.gyan.dev

**Reply by pfuller88** [reply]
silly question, which log file do I get the output from slingbox_server.  I'm on linux ubuntu.

**Reply by GerryDazoo** [reply]
have a look at your sling.service file.

**Reply by pfuller88** [reply]
I changed the py file to debug mode.  Nothing is getting through.   I changed the FFMPEG to your version.  I changed the xteve to the FFplay.exe and your parameters in the FFMPEG input option line and if failed with

FFMPEG log:             usage: ffplay [options] input_file
2022/11/19 13:50:56 [xTeVe] FFMPEG log:
2022/11/19 13:50:56 [xTeVe] FFMPEG log:             An input file must be specified
2022/11/19 13:50:56 [xTeVe] FFMPEG log:             Use -h to get full help or, even better, run 'man ffplay'

I also tried FFMPEG with the option from the earlier post.  FFMPEG fails with     FFMPEG error (Streaming was stopped by third party transcoder (FFmpeg / VLC))

Just cannot get this to work.  Thinking of giving up.

**Reply by pfuller88** [reply]
Ha, I got it working, finally.   I had to change the user agent setting in xteve from xTeVe to user_agent.  I read about at this link...
[https://github.com/xteve-project/xTeVe/issues/227](url)

**Reply by GerryDazoo** [reply]
What "debug mode" are you referring to?

**Reply by rbf121** [reply]
> Ha, I got it working, finally. I had to change the user agent setting in xteve from xTeVe to user_agent. I read about at this link... [https://github.com/xteve-project/xTeVe/issues/227](url)

Wow nice find.  I didn't have any issues with the default User Agent setting in xTeVe.  I wonder if this is a Windows only thing?  I'm running in docker (linux).

**Reply by pfuller88** [reply]
I tried both windows and linux,  both were failing until I changed the user agent.  Both now work with this change.   I  now have it running really well with the linux xteve running as a service.    I also changed the my m3u format as shown below, it picks up nicely in Plex and a little easier to map for the guide.

As for debug mode, I found in the last few lines of code in the slingbox_server.py file a "debug:False".  Changed to "debug:True".   More info dumped to the log file.

#EXTM3U
#EXTINF:-1 tvg-chno="895" tvg-id="NewsNation.895" tvg-logo="" group-title="News",NewsNation 895
http://127.0.0.1:8180/condo/500high?channel=895
#EXTM3U
#EXTINF:-1 tvg-chno="897" tvg-id="TheMovieChannel.897" tvg-logo="" group-title="Movie",The Movie Channel 897
http://127.0.0.1:8180/condo/500high?channel=897

**Reply by pfuller88** [reply]
Another note.  I had trouble with plex connecting and lagging a little.  I removed the check from "Disable video stream transcoding" in the Transcoder section in Plex.  It seems to work better for me with Plex having this flexibility.

**Reply by flipper2222222** [reply]
> > Ha, I got it working, finally. I had to change the user agent setting in xteve from xTeVe to user_agent. I read about at this link... [https://github.com/xteve-project/xTeVe/issues/227](url)
> 
> Wow nice find. I didn't have any issues with the default User Agent setting in xTeVe. I wonder if this is a Windows only thing? I'm running in docker (linux).

I am using windows and my user agent is  xTeVe

**Reply by GerryDazoo** [reply]
pfuller88: That debug flag is only for the Flask module that handles the web server for the remote. It has no effect on my pyhton code. I turned it off because it's output was only confusing people.

**Reply by pfuller88** [reply]
It help me see if I was sending codes to slinger via plex.

**Reply by pfuller88** [reply]
One other note for others on this setup.  I found resetting the xteve json to start numbering at 1 or another number caused a little problem when the number overlaps with my channel numbers and I have gaps in my channel numbers.   The xteve mapping will not overwrite a number, if you try to change a number to coincide with your channel number and that number is already in the xteve default list, xteve automatically changes your input to the next available number without warning.  I found it best to leave the default at 1000 and just type all my changes as my channel numbers had lots of gaps in the sequence.

**Reply by rbf121** [reply]
> I tried both windows and linux, both were failing until I changed the user agent. Both now work with this change. I now have it running really well with the linux xteve running as a service. I also changed the my m3u format as shown below, it picks up nicely in Plex and a little easier to map for the guide.
> 
> As for debug mode, I found in the last few lines of code in the slingbox_server.py file a "debug:False". Changed to "debug:True". More info dumped to the log file.
> 
> #EXTM3U #EXTINF:-1 tvg-chno="895" tvg-id="NewsNation.895" tvg-logo="" group-title="News",NewsNation 895 http://127.0.0.1:8180/condo/500high?channel=895 #EXTM3U #EXTINF:-1 tvg-chno="897" tvg-id="TheMovieChannel.897" tvg-logo="" group-title="Movie",The Movie Channel 897 http://127.0.0.1:8180/condo/500high?channel=897

I have a script I was working on to create the m3u file automatically as part of Gerry's slinger code.  I was thinking about including an optional file to map the channel numbers to the channel name like you have it.  I'll let you know when it's ready if you want to try it out.

**Reply by rbf121** [reply]
> One other note for others on this setup. I found resetting the xteve json to start numbering at 1 or another number caused a little problem when the number overlaps with my channel numbers and I have gaps in my channel numbers. The xteve mapping will not overwrite a number, if you try to change a number to coincide with your channel number and that number is already in the xteve default list, xteve automatically changes your input to the next available number without warning. I found it best to leave the default at 1000 and just type all my changes as my channel numbers had lots of gaps in the sequence.

Yea, this was an issue.  Luckily all my channels were mostly sequential so plex automatically assigned everything for me and it was very little work but if your channels are all spread out then you either still include the unused channels to have an easier time mapping sequentially in xteve/plex, or you manually map them correctly in xteve so Plex picks them up.  Or finally, you could attempt at applying some script or a regular expression to replace to go and "fix" all of the xteve mappings to match what your m3u file has.  

You could also download the xteve go code and modify that for your purposes, I messed around with it a bit to change the video/audio codec to display correctly in plex (Plex incorrectly lists the slingbox bitate as 40 mbps so I was trying to fix that, it didn't make any difference) and I changed some of the base configs to use the correct settings from the start but I never committed any changes to my fork. 

Let us know if you come up with a better way to do it but I just went path of least resistance and mapped channels 1-360.

**Reply by pfuller88** [reply]
it did not take me long to generate the 200 line m3u file.  i used excel to build the first line then just copied the formula down.  I cut and paste the channel lineup from my cable company’s website.

**Reply by pfuller88** [reply]
I'll also note that you can use the below website to further edit your m3u list.  For example, you upload your full list and then can edit it (use the "editor" button in the dropdown found in the top right) to screen out channels then download the results.  It is a free service, you will need to create an account.   [https://[m3u4u](https://m3u4u.com/).com/](url)

**Reply by Toei79** [reply]
Hey wondering its anyway works with pipe , ffmpg, I'm been trying use on tvheadend  but get errors while I play it, I feel.its close for what you are trying here , I need read all I guess.

My only goal its play it on tivimate but not works with explorer, but vlc its fine 


There its some script on github for you take epg from zap2it, I'm able to get ota and directv guides from there



edit checking i cant get it work

pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i http://192.168.2.131:8080/Gerry/450 -c copy -bsf:v h264_metadata=sample_aspect_ratio=4/3 -f mpegts -tune zerolatency pipe:1?



pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i http://192.168.2.131:8080/Gerry/450 -c copy -bsf:v h264_mp4toannexb -f mpegts -tune zerolatency pipe:1

pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i http://192.168.2.131:8080/Gerry/450 -c copy -bsf:v h264_metadata=sample_aspect_ratio=4/3 -f mpegts -tune zerolatency pipe:1?

pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i http://192.168.2.131:8080/Gerry/450 -c copy -f mpegts -tune zerolatency pipe:1

pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i http://192.168.2.131:8080/Gerry/450 -c copy -bsf:v h264_metadata=sample_aspect_ratio=4/3 -f mpegts -tune zerolatency pipe:1

i try those in tvheadend , i guess its the right ones for use there?

**Reply by rbf121** [reply]
Sorry I'm not familiar with tvheadend or tivimate but I assume it should work similar to what Xteve is doing.  Doubt anyone else has tried that integration so I think you will just need to experiment to get it to work properly.  Perhaps if you share the error message, someone might be able to help.

Regarding zap2it EPG data, is that a free service?  I have plex pass lifetime so I'm just using that but I did experiment with Jellyfin M3U Tuner using Schedules Direct for EPG but it didn't seem to pull in image content with the EPG.

**Reply by Toei79** [reply]
zap2it its online service, there its two scripts that i see working extracting the epg from there.
one of those its being running forever maybe like 10 years and working. 
i just add to my tvheadend and its fits to ota channels.  the directv i use to fill other channels.  format its .xml.

idk about that its image from shows? 
tivimate its kinda like the plex i guess but just tv channels. 
i dont think you get pics from shows , but logo channels yes. 

ok thank you , i just see you get kinda same error or something because tivimate just ''give up '' and give a parser error, 
tivimate read .m3u playlist , so i was thinking its be same think.

i try all those that you see there , that i see you use from there sadly not working.

**Comment by Toei79** [comment]
well i installed xteve and its running.  plus works on tivimate.

but then why not get working outside of xteve
pipe:///usr/bin/ffmpeg -hide_banner -loglevel error -fflags +genpts -i url-c copy -f mpegts -tune zerolatency pipe:1
i trieed this on tvheadend but no dice


if its used by xteve totally fine? its just need add buffer 0.5 or not?
